src/chain_spec.rs

```diff
@@ -39,23 +39,90 @@ use alloc::string::String;
 mod light_sync_state;
 mod structs;
 
+use crate::chain::chain_information::{ChainInformation, ChainInformationConsensus, BabeEpochInformation};
+
+pub struct LightSyncState {
+    inner: light_sync_state::DecodedLightSyncState,
+}
+
+fn convert_epoch(epoch: &light_sync_state::BabeEpoch) -> BabeEpochInformation {
+    let epoch_authorities: Vec<_> = epoch.authorities.iter()
+        .map(|authority| {
+            crate::header::BabeAuthority {
+                public_key: authority.public_key,
+                weight: authority.weight,
+            }
+        })
+        .collect();
+    
+    BabeEpochInformation {
+        epoch_index: epoch.epoch_index,
+        start_slot_number: Some(epoch.slot_number),
+        authorities: epoch_authorities,
+        randomness: epoch.randomness,
+        c: epoch.config.c,
+        allowed_slots: epoch.config.allowed_slots,
+    }
+}
+
+impl LightSyncState {
+    pub fn as_chain_information(&self) -> ChainInformation {
+        // Create a sorted list of all regular epochs that haven't been pruned from the sync state.
+        let mut epochs: Vec<_> = self.inner.babe_epoch_changes.epochs.iter()
+            .filter_map(|((_, block_num), epoch)| {
+                match epoch {
+                    light_sync_state::PersistedEpoch::Regular(epoch) => Some((block_num, epoch)),
+                    _ => None
+                }
+            })
+            .collect();
+
+        epochs.sort_unstable_by_key(|(&block_num, _)| block_num);
+
+        // Get the latest two epochs.
+        let prev_epoch = &epochs[epochs.len() - 2].1;
+        let current_epoch = &epochs[epochs.len() - 1].1;
```

Comment:
I still don't get how these epochs work, honestly. I thought that at least one epoch would start in the future relative to the finalized header, but in [polkadot-2251297.json](https://github.com/paritytech/substrate-lite/blob/main/bin/polkadot-2251297.json) the finalized header is for block `2251297`, `current_epoch` is for `2248920` and `prev_epoch` is for `2246525`.

Reply:
It might be that `babe_epoch_changes` contains the block number that contains the information about the epoch, and not the block number where the epoch starts.

Reply:
In other words, `prev_epoch` (value `2246525`) contains the information about the epoch that starts at block `2248920`, and `current_epoch` (value `2248920`) starts around block `226000`.

At least that's my guess.

Reply:
Since you can't know in advance at which block number an epoch is going to start, this is indeed the only thing that makes sense.